nova/src/circuits/supernova/mod.rs

```diff
@@ -0,0 +1,631 @@
+//! SuperNova implementation with CycleFold.
+//!
+//! Unlike [`IVCProof`](super::nova::sequential::IVCProof) allows the step circuit to have different
+//! structure based on the computed program counter.
+
+use std::marker::PhantomData;
+
+use ark_crypto_primitives::sponge::{
+    constraints::{CryptographicSpongeVar, SpongeWithGadget},
+    Absorb, CryptographicSponge,
+};
+use ark_ec::short_weierstrass::{Projective, SWCurveConfig};
+use ark_ff::{AdditiveGroup, BigInteger, PrimeField};
+use ark_r1cs_std::{fields::fp::FpVar, R1CSVar};
+use ark_relations::r1cs::{ConstraintSystem, ConstraintSystemRef, SynthesisError, SynthesisMode};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+
+use crate::{
+    absorb::CryptographicSpongeExt,
+    commitment::CommitmentScheme,
+    folding::nova::cyclefold::{
+        self,
+        nimfs::{
+            NIMFSProof, R1CSInstance, R1CSShape, R1CSWitness, RelaxedR1CSInstance,
+            RelaxedR1CSWitness,
+        },
+    },
+};
+use augmented::{
+    SuperNovaAugmentedCircuit, SuperNovaAugmentedCircuitInput,
+    SuperNovaAugmentedCircuitNonBaseInput,
+};
+
+pub mod public_params;
+pub use crate::folding::nova::cyclefold::Error;
+
+mod augmented;
+
+trait SuperNovaConstraintSynthesizer<F: PrimeField> {
+    fn generate_constraints(
+        self,
+        cs: ConstraintSystemRef<F>,
+    ) -> Result<(FpVar<F>, Vec<FpVar<F>>), SynthesisError>;
+}
+
+pub trait NonUniformCircuit<F: PrimeField>: Send + Sync {
+    /// Number of input variables, corresponds to `z.len()`.
+    const ARITY: usize;
+
+    /// Number of computable functions {F_0, ... , F_{l - 1}}.
+    const NUM_CIRCUITS: usize;
+
+    /// Return program counter for the first step.
+    fn pc_init(&self) -> u64 {
```

Comment:
What you suggest is default behavior, I don't see why it's not possible to start with the second augmented circuit or any other.

Reply:
i don't mean first among the set of instructions, i mean the first step in time. the initial `pc` is a function of `z_0`. We require that `F_j(z_0) = (_z, pc_init)` with `pc_init` the same for each choice of `j`. The paper writes this as `pc_init = phi(z_0); z_1 = F_(pc_init)(z_0)`. We could introduce `phi` in the code explicitly, or just note this stipulation.

Reply:
1. This would require implementors to be additionally careful with `NonUniformCircuit` trait.

2. `phi` can only be executed as part of some augmented circuit (say we don't care which one because of \all j requirement)
Does it mean that in the base case you first compute instanse `u'` that outputs `pc_init`, and then pass it as an input to the actual base step to get `u`? Or do you discard `u'`?

Reply:
I suggest something like 
```
pub trait NonUniformCircuit<F: PrimeField>: Send + Sync {
    /// Number of input variables, corresponds to `z.len()`.
    const ARITY: usize;

    /// Number of computable functions {F_0, ... , F_{l - 1}}.
    const NUM_CIRCUITS: u64;

    /// Generate constraints for computing the selector function "phi".
    ///
    /// Return the next value of the program counter.
    fn generate_constraints_pc(
        &self,
        cs: ConstraintSystemRef<F>,
        i: &FpVar<F>,
        z: &[FpVar<F>],
    ) -> Result<FpVar<F>, SynthesisError>;

    /// Generate constraints for applying function `F_(pc_value)` to `z`.
    ///
    /// Return the next value of `z`, `z_(i+1)`.
    fn generate_constraints_step(
        &self,
        cs: ConstraintSystemRef<F>,
        pc_value: u64,
        i: &FpVar<F>,
        z: &[FpVar<F>],
    ) -> Result<Vec<FpVar<F>>, SynthesisError>;

    /// Generate constraints for applying both `phi` and `F_{pc_value}` to `z`.
    ///
    /// Returns the next value of the program counter and the next value of `z`, `z_(i+1)`.
    /// The next value of the program counter is guaranteed not to depend on `pc_value`.
    fn generate_constraints(
        &self,
        cs: ConstraintSystemRef<F>,
        pc_value: u64,
        i: &FpVar<F>,
        z: &[FpVar<F>],
    ) -> Result<(FpVar<F>, Vec<FpVar<F>>), SynthesisError> {
        let pc = self.generate_constraints_pc(cs, i, z)?;
        let z_next = self.generate_constraints_step(cs, pc_value, i, z)?;
        Ok((pc, z_next))
    }
}
```

then in the `SupernovaAugmentedCircuit::generate_constraints()` you could have something like

``` 
    let pc = <SC as NonUniformCircuit>::generate_constraints_pc(circuit, cs, i, z)?;
    let pc_value: u64 = pc.value().into().unwrap();
    let z_next = <SC as NonUniformCircuit>::generate_constraints_step(circuit, cs, pc_value, i, z)?;
```
the relationship between `pc` and `pc_value` is _not_ enforced by the circuit, but just by the code. 

I think this is cleanest, but to be closer to the paper, you could have `generate_constraints()` take `pc_value` as input. then in a proof step, the prover first computes `pc_value = <SC as NonUniformCircuit>::generate_constraints_pc(circuit, dummy_cs, i, z)`, throwing out the constraint system/instance/witness, then uses this in the call to `generate_constraints()`.